source/_posts/2020-07-28-Kubernetes-controllers-A-New-Hope.md

```diff
@@ -0,0 +1,297 @@
+
+Today in Golang, when you compile to Wasm, all the standard library I/O operations goes through the [`syscall/js` module](https://golang.org/pkg/syscall/js/), which invokes a predefined set of imports in your module. From the doc: _Package js gives access to the WebAssembly host environment when using the js/wasm architecture. Its API is based on JavaScript semantics._ Going through the documentation should give you the idea of the purpose of this module:
+
+* [`Global()`](https://golang.org/pkg/syscall/js/#Global) allows you to get the `global` object
+* [`Invoke()`](https://golang.org/pkg/syscall/js/#Value.Invoke) allows you to use a `js.Value` as a function and invoke it
+
+Now let's analyze how the `net/http` client implementation on Wasm: When the user invokes `RoundTrip`, the request [is transformed in a Javascript object](https://github.com/golang/go/blob/b56791cdea5caa87ffcd585d29c294bd3d08a06a/src/net/http/roundtrip_js.go#L60), then using the syscall `Global` and `Call` the javascript global function [`fetch` is invoked](https://github.com/golang/go/blob/b56791cdea5caa87ffcd585d29c294bd3d08a06a/src/net/http/roundtrip_js.go#L106) and finally, on the returned promise, the [two callbacks are set](https://github.com/golang/go/blob/b56791cdea5caa87ffcd585d29c294bd3d08a06a/src/net/http/roundtrip_js.go#L167) to handle errors and responses.
+
+The problem with this approach, although it works well in the browsers, is that the **engine needs to implement all these JS functions used by Golang standard library**. Somehow, the engine should fake itself as a Javascript engine implementing the Javascript standard library. Not only, it requires to **secure access** to the _global_ object fields, in order to retain the isolation properties. Making `syscall/js` optional could make far more doable implementing controllers in Golang using Wasm.
+
+To make things even worse, at the moment there is no way to define imports and exports within the module, so we cannot define and evolve an ABI that fits our needs. The only doable solution with the existing tools consists in define "fake" Javascript functions on the host side for the imports and setting "fake" fields on the ["fake" global object for the exports](https://github.com/golang/go/issues/25612#issuecomment-506307400), but this approach makes the code far more complex and convoluted, both on host and Wasm module side.
+
+{% note info %}
+While the official Golang distribution [doesn't support](https://github.com/golang/go/issues/25612) definining imports and exports, [TinyGo supports it](https://github.com/tinygo-org/tinygo/tree/master/src/examples/wasm)
+{% endnote %}
+
+Markus managed to run the Memcached operator example of operator-sdk, using NodeJS as a host, hacking the [wasm_exec.js](https://github.com/golang/go/blob/master/misc/wasm/wasm_exec.js) to fix some polyfills. The final module ABI is:
```

Comment:
Not true, I ran a different one and it didn't successfully run just yet. Maybe keep this as-is though, i'll retry.

I also didn't hack wasm_exec.js but had to load a polyfill for `fetch` and hack it to conform to browsers APIs.

Reply:
Ok i'll try to rephrase a bit and then let me know your results, i wanna log them here (maybe just update directly this post with your results)

Reply:
look at the reword now